dataplane/handlers/interface.go

```diff
@@ -0,0 +1,293 @@
+type Interface struct {
+	c             *ygnmi.Client
+	watchCancelFn context.CancelFunc
+	linkDoneCh    chan struct{}
+	addrDoneCh    chan struct{}
+	fwd           fwdpb.ServiceClient
+	mu            sync.RWMutex
+	state         map[string]*oc.Interface
+	idxToName     map[int]string
+}
+
+// NewInterface creates a new interface handler.
+func NewInterface(yc *ygnmi.Client, fwd fwdpb.ServiceClient) *Interface {
+	return &Interface{
+		c:         yc,
+		fwd:       fwd,
+		idxToName: map[int]string{},
+		state:     map[string]*oc.Interface{},
+	}
+}
+
+// Start starts running the handler, watching the cache and the kernel interfaces.
+func (ni *Interface) Start(ctx context.Context) error {
+	log.Info("starting interface handler")
+	b := &ocpath.Batch{}
+
+	if err := ni.setupPorts(ctx); err != nil {
+		return fmt.Errorf("failed to setup ports: %v", err)
+	}
+
+	b.AddPaths(
+		ocpath.Root().InterfaceAny().Enabled().Config().PathStruct(),
+		ocpath.Root().InterfaceAny().Name().Config().PathStruct(),
+		ocpath.Root().InterfaceAny().Ethernet().MacAddress().Config().PathStruct(),
+		ocpath.Root().InterfaceAny().Subinterface(0).Ipv4().AddressAny().Ip().Config().PathStruct(),
+		ocpath.Root().InterfaceAny().Subinterface(0).Ipv4().AddressAny().PrefixLength().Config().PathStruct(),
+		ocpath.Root().InterfaceAny().Subinterface(0).Ipv6().AddressAny().Ip().Config().PathStruct(),
+		ocpath.Root().InterfaceAny().Subinterface(0).Ipv6().AddressAny().PrefixLength().Config().PathStruct(),
+	)
+	cancelCtx, cancelFn := context.WithCancel(ctx)
+	ni.watchCancelFn = cancelFn
+
+	watcher := ygnmi.Watch(cancelCtx, ni.c, b.Config(), func(val *ygnmi.Value[*oc.Root]) error {
+		log.Info("reconciling interfaces")
+		root, ok := val.Val()
+		if !ok || root.Interface == nil {
+			return ygnmi.Continue
+		}
+		for _, i := range root.Interface {
+			ni.reconcile(i)
+		}
+		return ygnmi.Continue
+	})
+
+	ni.linkDoneCh = make(chan struct{})
+	linkUpdateCh := make(chan netlink.LinkUpdate)
+	ni.addrDoneCh = make(chan struct{})
+	addrUpdateCh := make(chan netlink.AddrUpdate)
+
+	if err := netlink.LinkSubscribe(linkUpdateCh, ni.linkDoneCh); err != nil {
+		return fmt.Errorf("failed to sub to link: %v", err)
+	}
+	if err := netlink.AddrSubscribe(addrUpdateCh, ni.addrDoneCh); err != nil {
+		return fmt.Errorf("failed to sub to addr: %v", err)
+	}
+
+	go func() {
+		for {
+			select {
+			case up := <-linkUpdateCh:
+				ni.handleLinkUpdate(&up)
+			case up := <-addrUpdateCh:
+				ni.handleAddrUpdate(&up)
+			}
+		}
+	}()
+
+	go func() {
+		// TODO: handle error
+		if _, err := watcher.Await(); err != nil {
+			log.Warningf("interface watch err: %v", err)
+		}
+	}()
+
+	return nil
+}
+
+// Stop stops all watchers.
+func (ni *Interface) Stop() {
+	ni.watchCancelFn()
+	close(ni.linkDoneCh)
+	close(ni.addrDoneCh)
+}
+
+// reconcile compares the interface config with state and modifies state to match config.
+func (ni *Interface) reconcile(config *oc.Interface) {
```

Comment:
(no action required) I like this pattern: it cleanly separates out the reconciliation action from the watch action.